### ISSUE TITLE
create chart testing framework

### ISSUE THREAD

**Issue by michelleN** [kubernetes/helm#1163]
once you've deployed a chart, run a series of tests against your release to ensure expected outcome

**Comment by thomastaylor312** [comment]
@michelleN: Do you have an overview of how all of this is going to work? I'd like to see what it is going to do and provide feedback/help if desired 😄

**Comment by michelleN** [comment]
Hey @thomastaylor312 - would love any thoughts or feedback. Below is an overview.

**The problem:**
	Chart consumers may not always know the chart author’s intention. Therefore, when a chart consumer installs a chart, he or she needs to be able to verify that the installation did exactly what it was supposed to do. In other words, the the application works as intended and the configuration specified in the values file and templating logic gives the desired outcome.

**The solution:**
	We want chart authors to be able to specify definitions of success in the form of tests and package those with a chart. We also want chart consumers to be able run these tests to validate that the functionality of the chart is what the author desired.

_**Example questions that running a test might answer:**_
Does my username and password that I specified for my database actually work? Do the wrong credentials not work?
Does my web service work correctly?
I specified a service in my chart that was supposed to create a load balancer. Did that happen?
Is there data in my database after an upgrade?

**High level implementation of an MVP:**
We’re currently working on a `helm test <release>` command. It will fetch a release, run the tests, and return the results. A test is a pod definition with a hook annotation (`helm.sh/hook: test`) that specifies a container that runs a particular command. If that pod is in the completed state and the container exits 0, then we consider that a successful run. Any non-zero exit is considered a failure. This lets you abstract what your test is from the framework in helm that lets you run the test. 

A chart author may specify multiple tests in a chart. We refer to the entire set of tests in a chart as a `test suite`. `helm status` will display the results from the last run of the test suite.

**Comment by michelleN** [comment]
#1777 is a WIP PR for chart testing. You are welcome to check that out, build it. After installing the chart, you can run the tests using the `helm test` command to get an idea for the feel.

**Comment by thomastaylor312** [comment]
@michelleN Thanks for the good overview! So here are a few points of feedback/ideas:

So I feel like there are two different behaviors people would want here:
1. Tests - Ensuring that all cases are handled properly (kind of like unit tests). 
2. Validation - Making sure the application is up

**Tests**
These would test things that a specific user might not care about, but are important to make sure the Chart will work in every way it is intended too. So if I have a nginx server that I want to work with 10 different configurations, I would want to deploy the Chart 10 times and validate it is working each time before cleaning up.

**Validation**
These would take care of making sure things are up and running as described in your overview. With the current behavior as described, I feel that instead of a separate command, this would be more along the lines of a flag called `--validate` on an `upgrade`, `rollback`, or `install`. Because if I request validation and it fails, I would want the release to be marked as failed.

I know my idea of "tests" might be out of scope for what you are trying to do here, but I wanted to bring up the overlap with your idea. Also, this idea of validation may overlap somewhat with I was working on with #1805 

Thoughts? I know this is a lot.

**Comment by michelleN** [comment]
On Tests:
Your idea of tests are really interesting especially because it's from the chart author's perspective. It would also be helpful for CI on the charts repo to be able to run different configurations and validate that the chart is working for a given matrix of configurations the way its supposed to.
I think what we're building right now in #1777 will set the foundation for describing how to validate a chart with a given configuration. Does what you are suggesting mean that a testing framework would also need to be able to consume different values files?
I'm interested in hearing more about how a chart author would describe the different configurations.


On Validation:
I agree that a flag makes sense here. What makes this automation difficult is that we don't really know exactly what it means for each resource type to actually be available and then there are the other requirements of making sure a PVC is bound or an external load balancer is up and running if it's specified. For an MVP, we chose not to bite off this part but instead focus on how to express a test definition and run a test definition and query test results. 
I think your work #1805 could really help here and we could integrate it into the test framework. I'd be interested in chatting more about this and seeing what bumps, if any, you run into while solving #1805.

**Comment by thomastaylor312** [comment]
I completely agree that this work you are doing now would lay the foundation for tests. As for describing different configurations, I was thinking through stuff, but I didn't come to a complete answer. The best one I thought of was to have a `test` directory in a chart that would have filenames like `{thing I am testing for}.yaml` that would provide the specific sets of values you test against. Then the `helm test` command would know to look there by default (or another directory the user specifies) and then loop over the values files to do its tests.

I am hoping to have a PR open for #1805 today, so I'll link to this issue when I do.

**Comment by thomastaylor312** [comment]
@michelleN: The `--wait` [PR](https://github.com/kubernetes/helm/pull/1820)

**Comment by nrvale0** [comment]
I'm joining this conversation late but I arrived here via conversation in the #helm Slack channel. Generally, I like the direction of this conversation and hope I might be able to contribute in some way.

As to the workflow from a Chart authors perspective, it might be useful to familiarize with the approach(es) to enabling this sort of pre-deployment validation and testing (in the Puppet world we called in Unit and Acceptance testing) in tools like Chef's InSpec, infrataster, serverspec, and other rspec-based tooling. There are similar, though fewer, tools in the Python world (TestInfra) and I've started looking for similar things from the Golang community without a lot of success so far. I mention various languages only because it might help to direct exploration of the underlying ideas and not to start any sort of language war. The hook-pointing-to-sidecar pattern above actually seems like it would be very non-opinionated as to language/toolkit.

**Comment by rothgar** [comment]
I'm also late to the conversation (brought in via slack) but I'm curious if tests can be run pre-deployment.

This would be useful when charts make assumptions about a cluster
e.g.
- dynamic disk provisioner available
- loadBalancer available for service

I think charts should make as few assumptions about a cluster as possible but should present data to the user before deploying to let the user know if they need to update the values.yaml file or if a chart simply won't work because their cluster setup is incompatible (old version of kubernetes)

A real world example that just happened for a friend was he was trying to deploy stable/redis and it wouldn't start because [persistence was enabled by default](https://github.com/kubernetes/charts/blob/master/stable/redis/values.yaml#L22) which assumed an alpha feature (dynamic provisioning) was available when it wasn't. Settings like this should be disabled by default but let the user know what the default setting is and possibly how it may impact them. e.g. Your mariadb pod is not using persistent storage and will lose all data when the pod restarts.

**Comment by foxish** [comment]
@michelleN, any update on this issue? Is it blocked on something else?

**Comment by michelleN** [comment]
@foxish no blockers. the PR is in review

**Comment by technosophos** [comment]
At what point do we mark this issue as closed?

**Comment by thomastaylor312** [comment]
@technosophos Do we want to split the validation idea out into its own issue?

**Comment by michelleN** [comment]
@technosophos this can be closed. All follow up tasks in separate issues already.

**Comment by michelleN** [comment]
+1 to splitting validation idea out into its own issue

**Comment by technosophos** [comment]
Closing this as resolved. YAY!

**Comment by thomastaylor312** [comment]
Validation [issue](https://github.com/kubernetes/helm/issues/1952)

**Comment by goginenigvk** [comment]
Is there any validation script that can ensure  the application is actually running properly.
